### ISSUE TITLE
Video freezes instantly on UWP

### ISSUE THREAD

**Issue by cy238** [microsoft/MixedReality-WebRTC#94]
Hello, I'm currently trying to replace an old version of webrtc-uwp-sdk with this library in a HoloLens project and encountered some issues.

I have a very simple test project (basically copied from the Unity sample in this repo) with custom signaling. Everything works great in the editor and in standalone builds.
On UWP as soon as the connection is fully established the video feed of the UWP client just freezes. Audio keeps working in both directions and the video feed from standalone to UWP works aswell.
As far as I can tell the custom signaling does not cause this, since it also happens with NodeDSS.

There are log entries related to the H264 decoder exactly when the freeze occurs, but I couldn't find out what triggers them:
```
WinUWPH264DecoderImpl::Release()
WinUWPH264DecoderImpl::~WinUWPH264DecoderImpl()
WinUWPH264DecoderImpl::Release()
```

I tried using different video codecs for both clients. Setting VP8 for both clients still causes the same behaviour. Setting H264 for both clients causes the standalone build to crash with this exception:
```
ArgumentException: Value does not fall within the expected range.
Microsoft.MixedReality.WebRTC.Interop.Utils.ThrowOnErrorCode (System.UInt32 res) (at D:/Dev/MR-WebRTC/libs/Microsoft.MixedReality.WebRTC/Interop/InteropUtils.cs:117)
Microsoft.MixedReality.WebRTC.PeerConnection.OnLocalSdpReadytoSend (System.String type, System.String sdp) (at D:/Dev/MR-WebRTC/libs/Microsoft.MixedReality.WebRTC/PeerConnection.cs:1315)
Microsoft.MixedReality.WebRTC.Interop.PeerConnectionInterop.LocalSdpReadytoSendCallback (System.IntPtr userData, System.String type, System.String sdp) (at D:/Dev/MR-WebRTC/libs/Microsoft.MixedReality.WebRTC/Interop/PeerConnectionInterop.cs:148)
UnityEngine.UnhandledExceptionHandler:<RegisterUECatcher>m__0(Object, UnhandledExceptionEventArgs)
```

Any ideas?

**Comment by djee-ms** [comment]
Hi @cy238,

There is currently a bug with the UWP video encoder (#74). However this is I think specific to H264, and the fact that it also fails for you with VP8 means this is unrelated since VP8 is a software encoder with a very different codepath as H264 on UWP. But mentioning it just in case...

For the argument exception you get, the call to `SdpForceCodecs()` is failing. This means that the internal SDP message buffer is too small. This is quite unexpected, because if you filter a codec, then you are removing some information from the SDP message, so the filtered message should always be at most equal, but generally smaller than the input one. I have no idea why this happens. If you put a breakpoint on that line, after the call to `Utils.SdpForceCodecs()` returned:
```cs
Utils.ThrowOnErrorCode(res);
```
then what is the value of `lengthInOut`? It should be <= `builder.Capacity`, but if the exception is thrown I am guessing it isn't. What are those two values?

For the rest, I have no idea at this point. In Debug configuration you should be able to see more logs from WebRTC; could you please attach them to this issue? If the decoder is released, it sounds like the video is closing, but that doesn't explain why.

**Comment by cy238** [comment]
Hi @djee-ms,

after I read your comment about the uwp-webrtc-sdk decoder for H264 only being available on UWP in #88, I switched both clients to UWP (which was our plan for the project anyway) and it worked flawlessly with H264.

After that I went back to a standalone build to find out about the issue with `SdpForceCodecs()` and... well... it just worked without the ArgumentException. I honestly have no clue why as there were zero code changes. Probably the usual Unity weirdness...

The only remaining problem that I have is ending a "call" or resetting the connection. How would you handle that? My current idea is to use `GameObject.Destroy` on the PeerConnection and simply recreating it after a call ends. This way I wouldn't have to dig too deep into async initialization logic and event hookups. Does that seems like an okay solution?

**Comment by djee-ms** [comment]
Yes once the call is finished you need to tear down the `PeerConnection` and recreate a new one. I am pretty sure this is the safest for the C# library, and I think this is also the only way from the point of view of the underlying C++ implementation, as I don't recall seeing any indicator that Google's implementation allows reusing a native peer connection object (maybe it can, but I am not sure what cleaning needs to be done before that). So creating a new `PeerConnection` component in Unity sounds like the safest bet in your case.

**Comment by djee-ms** [comment]
@cy238 there was actually an off-by-one bug in `SdpForceCodecs()` when the filtering does not filter anything because the codec is not found, and the output message is equal to the input one, but with a null-terminator added with overflows past the input buffer capacity. Fixed in 9a2996d3ec2c286ae717f0ff9a84590eaa09ef74.
